---
title: "10_analysis"
author: "Victor Yuan"
date: "April 21, 2019"
output:
  html_document:
    keep_md: true
    toc: true
    toc_depth: 4
    toc_float:
      collapsed: false
    theme: spacelab
editor_options: 
  chunk_output_type: console
---

This is for analysing the scRNAseq data
# 1.0 Libraries and Data

```{r}
set.seed(2019)
library(readxl) # read data in
library(plyr) 
library(stringr)
library(knitr);library(kableExtra) # for displaying pretty tables
library(dplyr)
library(Seurat) # V3
library(readr)
library(ggplot2)
library(cowplot)
library(tidyr)
library(monocle)
library(slingshot)
library(RColorBrewer)
troph <- readRDS('../../data/interim/02_troph_merge.rds')
```

# 2.0 Validate hits

First subset to trophoblast

## 2.1 List of hits

Obtained from Microarray results folder

### Violin plots

```{r}
hitlist <- read_xlsx('../../data/microarray results/2019-03-19 currated scRNA-seq gene list.xlsx')
hitlist <- hitlist %>% gather(key = 'Geneset', value = 'Gene') %>% filter(!is.na(Gene))

#updated hit list:
top_20 <- read.csv('../../data/microarray results/topups_Oxygen20_1 2.csv') %>% as_tibble()
top_1 <- read.csv('../../data/microarray results/topdowns_Oxygen20_1.csv') %>% as_tibble()

x <- top_20 %>% slice(1:15) %>% select(SYMBOL) %>% mutate(Geneset = 'Top Upregulated (20%)')
y <- top_1%>% slice(1:15) %>% select(SYMBOL) %>% mutate(Geneset = 'Top Upregulated (1%)')
top <- bind_rows(x, y) %>% rename(Gene = SYMBOL) %>% select(Geneset, Gene)

# combine updated to old hit list
hitlist <- hitlist %>% filter(!Geneset %in% c('Target High', 'Target Low')) %>% bind_rows(top)

# order levels to plot by
Idents(troph) <- factor(as.character(troph$celltype), 
                        levels = c('VCT', 'VCT p', 'EVT', 'EVT p', 'SCT'))

#violin plots
VlnPlot(object = troph, features = hitlist$Gene[hitlist$Geneset=='Troph (general)'], 
        pt.size = F, ncol = 4, same.y.lims =  T) 
VlnPlot(object = troph, features = hitlist$Gene[hitlist$Geneset=='CTB'],
        pt.size = F, same.y.lims =  T)
VlnPlot(object = troph, features = hitlist$Gene[hitlist$Geneset=='Dist CCTB'], 
        pt.size = F, same.y.lims =  T)
VlnPlot(object = troph, features = hitlist$Gene[hitlist$Geneset=='Prox CCTB'],
        pt.size = F, same.y.lims =  T)
VlnPlot(object = troph, features = hitlist$Gene[hitlist$Geneset=='Syn Pre CTB'],
        pt.size = F, same.y.lims =  T)
VlnPlot(object = troph, features = hitlist$Gene[hitlist$Geneset=='Top Upregulated (20%)'],
        pt.size = F, same.y.lims =  T, ncol = 5)
VlnPlot(object = troph, features = hitlist$Gene[hitlist$Geneset=='Top Upregulated (1%)'],
        pt.size = F, same.y.lims =  T, ncol = 5)
```

### heatmap

```{r}
Idents(troph) <- 'celltype'
#plot average expression
average_exp <- AverageExpression(object = troph, return.seurat = T)
ave_exp_mat <- average_exp@assays$integrated@scale.data[,] %>% as.data.frame()

library(pheatmap)
#create annotation dataframe for genes
row_label <- as.data.frame(hitlist) 
row_label$Gene <- make.unique(row_label$Gene)
rownames(row_label) <- gsub('\\.', '-', row_label$Gene)
row_label$Geneset <- factor(row_label$Geneset, 
                            levels = c('Troph (general)', 'CTB', 'Dist CCTB', 'Prox CCTB', 
                                       'Syn Pre CTB', 'Top Upregulated (20%)', 'Top Upregulated (1%)'))

row_label <- row_label[intersect(rownames(row_label), rownames(ave_exp_mat)),]
ave_exp_mat <- ave_exp_mat[rownames(row_label),]

colors_heat <- list(
  Geneset = c('Troph (general)' = 'grey',
              'CTB' = '#FFEC23',
              'Dist CCTB' = '#E41A1C',
              'Prox CCTB' = '#FF8064',
              'Syn Pre CTB' = '#4DAF4A',
              'Top Upregulated (20%)' = '#00C9BF',
              'Top Upregulated (1%)' = '#984EA3')
)

pheatmap(ave_exp_mat[,c('VCT', 'VCT p', 'EVT', 'EVT p', 'SCT')], 
         cluster_rows = F, cluster_cols = F, 
         annotation_row = row_label[,'Geneset',F],
         annotation_colors = colors_heat,
         color = RColorBrewer::brewer.pal(9, "Blues"),
         cellwidth = 15, cellheight = 15)

colors_heat2 <- RColorBrewer::brewer.pal(7, 'Dark2')
names(colors_heat2) <- c('Troph (general)', 'CTB', 'Dist CCTB', 'Prox CCTB', 
                         'Syn Pre CTB', 'Top Upregulated (20%)', 'Top Upregulated (1%)')

heatmap_plot <- pheatmap(ave_exp_mat[,c('VCT', 'VCT p', 'EVT', 'EVT p', 'SCT')], 
         cluster_rows = F, cluster_cols = F, 
         annotation_row = row_label[,'Geneset',F],
         annotation_colors = list('Geneset' = colors_heat2),
         color = RColorBrewer::brewer.pal(9, "Blues"),
         cellwidth = 15, cellheight = 15)
heatmap_plot
```

pdf('../../outs/fig5c.pdf', height = 10)
heatmap_plot
dev.off()

# 3.0 Pseudotime

Note that because we are importing seurat merged data, we have to set the `expressionFamily` 
argument as 'gaussianff'. Then, in following functions, when possible we must specify the arguments
`norm_method` = 'none' and `pseudo_expr` = 0.

## 3.1 Import into monocle

```{r}
# import into monocle object
#Extract data, phenotype data, and feature data from the SeuratObject
troph_data <- as(as.matrix(troph@assays$integrated@data), 'sparseMatrix')

pd <- new('AnnotatedDataFrame', data = troph@meta.data)

fData <- data.frame(gene_short_name = row.names(troph_data), row.names = row.names(troph_data))
fd <- new('AnnotatedDataFrame', data = fData)

#Construct monocle object
monocle_cds <- newCellDataSet(troph_data,
                         phenoData = pd,
                         featureData = fd,
                         lowerDetectionLimit = 0.5,
                         expressionFamily = uninormal())

# process, not necessary unless using negbinomial 
#monocle_cds <- estimateSizeFactors(monocle_cds)
#monocle_cds <- estimateDispersions(monocle_cds) # not neessary for gaussian data

# remove genes with very low expression (noninformative)
#monocle_cds <- detectGenes(monocle_cds, min_expr = 0.1)

# sellect a minimum number of cells expressing a gene as a threshold
#fData(monocle_cds)$use_for_ordering <- 
#  fData(monocle_cds)$num_cells_expressed > 0.05 * ncol(monocle_cds)
```

## 3.2 Create trajectory

Here we try pseudotime analysis on the trophoblast cells using 3 different approaches to generate 
the trajectory.

### Ordering based on cluster DEGs

Here we are finding clusters on a reduced dimensional projection (PCA -> TSNE), then 
clustering ('density-peak'), then finding the top 1000 differentially expressed genes, then 
performing the pseudotime analysis on those genes. This is the recommended approach from the
documentation.

```{r}
# select number of principal components to use for pseudotime
plot_pc_variance_explained(cds = monocle_cds, max_components = 20, return_all = F, 
                           pseudo_expr = 0, norm_method = 'none') # required argumnets for gaussianff

set.seed(2019)
monocle_cds <- reduceDimension(monocle_cds, max_components = 2, 
                               norm_method = 'none', pseudo_expr = 0,
                               num_dim = 10, reduction_method = 'tSNE', verbose = T)

# run density peak clustering on reduced projection
monocle_cds <- clusterCells(monocle_cds, rho_threshold = 2,
                         delta_threshold = 10,
                         skip_rho_sigma = T,
                         verbose = F)

#  this plot can help us decide number of clusters
plot_rho_delta(monocle_cds, rho_threshold = 2, delta_threshold = 10)

plot_cell_clusters(monocle_cds, color_by = 'as.factor(Cluster)')
plot_cell_clusters(monocle_cds, color_by = 'as.factor(celltype)')
```

Run DEG on clusters

~45 minutes with 24 cores

```{r eval = F}
# find differentially expressed genes between all clusters
clustering_DEG_genes <- differentialGeneTest(monocle_cds,
                                             fullModelFormulaStr = '~Cluster',
                                             cores = 24)
```

saveRDS(clustering_DEG_genes, '../../data/interim/clustering_DEG_genes.rds')

```{r}
clustering_DEG_genes <- readRDS('../../data/interim/clustering_DEG_genes.rds')

# select top 1000 differnetially expressed genes to order the cells
ordering_genes <- row.names(clustering_DEG_genes)[order(clustering_DEG_genes$qval)][1:1000]
monocle_cds <- setOrderingFilter(monocle_cds, ordering_genes = ordering_genes)
monocle_cds <- reduceDimension(monocle_cds, method = 'DDRTree', 
                               norm_method = 'none', pseudo_expr = 0)
monocle_cds <- orderCells(monocle_cds)
plot_cell_trajectory(monocle_cds, color_by = "as.factor(celltype)")
plot_cell_trajectory(monocle_cds, color_by = "Cluster")
```

Pretty good.

### Based on top 1000 variable genes

```{r, eval = T}
top1000variable <- troph@assays$integrated@var.features[1:1000]
monocle_cds_top1000 <- setOrderingFilter(monocle_cds, ordering_genes = top1000variable)
monocle_cds_top1000 <- reduceDimension(monocle_cds_top1000, method = 'DDRTree',
                               norm_method = 'none', pseudo_expr = 0)
monocle_cds_top1000 <- orderCells(monocle_cds_top1000)
plot_cell_trajectory(monocle_cds_top1000, color_by = "as.factor(celltype)")

plot_cell_trajectory(monocle_cds_top1000, color_by = "Pseudotime")

```

This looks pree good. Can it be better?

### Based on top 1000 DEGs between all trophoblast subtypes

```{r eval = F}
# find differentially expressed genes between trophoblasts
troph_DEG_genes <- differentialGeneTest(monocle_cds,
                                        fullModelFormulaStr = '~celltype',
                                        cores = 24)
```

saveRDS(troph_DEG_genes, '../../data/interim/troph_DEG_genes.rds')

```{r, eval = T}
troph_DEG_genes <- readRDS('../../data/interim/troph_DEG_genes.rds')

ordering_genes <- row.names(troph_DEG_genes)[order(troph_DEG_genes$qval)][1:1000]
monocle_cds_DEG <- setOrderingFilter(monocle_cds, ordering_genes = ordering_genes)
monocle_cds_DEG <- reduceDimension(monocle_cds_DEG, method = 'DDRTree', 
                               norm_method = 'none', pseudo_expr = 0)
monocle_cds_DEG <- orderCells(monocle_cds_DEG)
plot_cell_trajectory(monocle_cds_DEG, color_by = "as.factor(celltype)")

```

Best so far

### Based on all genes

Takes > 12 hours, will omit for now.

```{r eval = T}
troph_expressed_genes <- rownames(troph)[Matrix::rowMeans(troph@assays$RNA@counts) > 0.1]
length(troph_expressed_genes) # 10491

ordering_genes <- troph_expressed_genes
monocle_cds_all <- setOrderingFilter(monocle_cds, ordering_genes = ordering_genes)
monocle_cds_all <- reduceDimension(monocle_cds_all, method = 'DDRTree', 
                               norm_method = 'none', pseudo_expr = 0)
monocle_cds_all <- orderCells(monocle_cds_all)
plot_cell_trajectory(monocle_cds_all, color_by = "as.factor(celltype)")
```

### Based on trophoblast marker genes.

First we assign cell type by trophoblast marker genes

```{r eval = T}
EGFR_id <- row.names(subset(fData(monocle_cds), gene_short_name == "EGFR"))
HLAG_id <- row.names(subset(fData(monocle_cds), gene_short_name == "HLA-G"))
ENDOU_id <- row.names(subset(fData(monocle_cds), gene_short_name == "ENDOU"))
MKI67_id <- row.names(subset(fData(monocle_cds), gene_short_name == "MKI67"))
  

cth <- newCellTypeHierarchy()

cth <- addCellType(cth,
                   cell_type_name = "CTB",
                   classify_func = function(x) { x[EGFR_id,] >= 0.1 })

cth <- addCellType(cth,
                   cell_type_name = "CTB p", parent_cell_type_name = "CTB",
                   classify_func = function(x) { x[MKI67_id,] >= 0.1 })

cth <- addCellType(cth,
                   cell_type_name = "EVT",
                   classify_func = function(x) { x[HLAG_id,] >= 0.1 })

cth <- addCellType(cth,
                   cell_type_name = "EVT p", parent_cell_type_name = "EVT",
                   classify_func = function(x) { x[MKI67_id,] >= 0.1 })

cth <- addCellType(cth,
                   cell_type_name = "STB",
                   classify_func = function(x) { x[ENDOU_id,] >= 0.1})

monocle_cds_class <- classifyCells(monocle_cds, cth)
```

Now we select additional genes that covary with these marker genes

```{r eval = T}
# do this on only the genes with expression over a minimum value
troph_expressed_genes <- rownames(troph)[Matrix::rowMeans(troph@assays$RNA@counts) > 0.1]
length(troph_expressed_genes) # 10491

# find expressed genes
marker_diff <- markerDiffTable(monocle_cds_class[,],
                               cth, cores = 24)
semisup_clustering_genes <-
    row.names(subset(marker_diff, qval < 0.05))
semisup_clustering_genes <-
    row.names(marker_diff)[order(marker_diff$qval)][1:1000]
```

```{r eval = T}
monocle_cds_class <- setOrderingFilter(monocle_cds, semisup_clustering_genes)
monocle_cds_class <- reduceDimension(monocle_cds_class, max_components = 2,
    method = 'DDRTree', norm_method = 'none', pseudo_expr = 0)
monocle_cds_class <- orderCells(monocle_cds_class)
plot_cell_trajectory(monocle_cds_class, color_by = "celltype") +
    theme(legend.position = "right")
```

## 3.3 Plot gene expression along trajectory

Start with overview (fig 5a, 5b)
### Fig5a, 5b

```{r}
cols1 <- c('#919191', '#feb24c', '#fd8d3c', '#fc4e2a', '#e31a1c')
cols2 <- c('#919191', '#919191', # greys
           '#ffeda0', '#feb24c', '#fc4e2a', # yellow/orange
           '#e31a1c', '#bd0026') # red
cols3<- c('#919191', '#919191', '#919191', # greys
          '#feb24c', '#fc4e2a', # yellow/orange
           '#e31a1c', '#bd0026') # red

# add expression data to pdata
pDat <- monocle_cds@phenoData@data %>% as_tibble %>%
  mutate(dim1 = reducedDimS(monocle_cds)[1,], dim2 = reducedDimS(monocle_cds)[2,]) %>%
  bind_cols(t(troph@assays$integrated@data[hitlist$Gene,]) %>% as.data.frame)

pDat_melt<- pDat %>% select(dim1:NRN1, celltype) %>% 
  gather(key = Gene, value = Expression, -dim1, -dim2, -celltype) %>%
  left_join(hitlist, by = c('Gene'))

# plot one gene test
ggplot(pDat_melt %>% filter(Gene == 'OAS1'), aes(x = dim1, y = dim2, col = Expression)) +
  geom_point() +
  scale_color_gradientn(colours = cols3)


# overview
fig5a <- ggplot(pDat, aes(x = dim1, y = dim2, col = celltype)) +
  geom_point(alpha = 0.7, size = 1) +
  scale_color_brewer(palette = 'Set1')+
  labs(x = 'Dimension 1', y = 'Dimension 2')
fig5a
  
  
# classic trophoblast markers
fig5b <- ggplot(pDat_melt %>% filter(Gene %in% c('KRT7', 'EGFR', 'ERVFRD-1', 'HLA-G')) %>%
         mutate(Gene = factor(as.character(Gene), levels = c('KRT7', 'EGFR', 'ERVFRD-1', 'HLA-G'))),
       aes(x = dim1, y = dim2, col = Expression)) +
  geom_point(size = 0.8, alpha = 0.8) +
  scale_color_gradientn(colours = cols2) +
  facet_wrap(~Gene) +
  labs(x = 'Dimension 1', y = 'Dimension 2')
fig5b
```

tiff(file = '../../outs/fig5a.tiff',width = 7.5, height = 5, units = 'in', res = 300)
fig5a
dev.off()

tiff(file = '../../outs/fig5b.tiff',width = 9.5, height = 7, units = 'in', res = 300)
fig5b
dev.off()

Validate O2 hits

###  1%
 
Upregulated in 1%

```{r}
#facet grid
fig5d <- ggplot(pDat_melt %>% filter(Geneset == 'Top Upregulated (1%)'),
                aes(x = dim1, y = dim2, col = Expression)) +
  geom_point(size = 0.8, alpha = 0.8) +
  scale_color_gradientn(colours = cols2) + 
  facet_wrap(~Gene, scale = 'free', nrow = 3) +
  labs(x = 'Dimension 1', y = 'Dimension 2')
fig5d


# with cowplot::plot_grid()
plot_list_low <- list()

gene_ind <- hitlist %>% filter(Geneset == 'Top Upregulated (1%)') %>% pull(Gene)
  
  #c('LOX', 'FOXO1', 'HEY1', 'HCAR3', 'LPCAT2', 'FLT1', 'HTRA1', 'EGLN3', 'CSF1R', 'PLAUR')
  
for (i in 1:length(gene_ind)) {
  if(gene_ind[i] %in% c('LOX', 'ACTC1', 'BIRC7', 'SPNS2', 'AK4')) {
    cols <-  c('#919191', # greys
            '#fc4e2a', # /orange
           '#e31a1c', '#bd0026') # red # high contrast color scale
  } else if (gene_ind[i] %in% c('TNFSF10', 'S100A4', 'JAM2', 'EGLN3')) {
    cols <- c('#919191', '#919191', '#919191', # greys
            '#fc4e2a', # /orange
           '#e31a1c', '#bd0026')
  } else if (gene_ind[i] %in% c('RORA', 'AK4', 'TSC22D3', 'NRN1')) {
    cols <- c('#919191', '#919191', '#919191', # greys
            '#FEB24C', '#fc4e2a', # /orange
           '#e31a1c', '#bd0026')
  } else {
    cols <- c('#919191', '#919191',  # greys
            '#fc4e2a', # /orange
           '#e31a1c', '#bd0026') # red
  }
  
  plot_list_low[[i]] <-
    ggplot(pDat_melt %>% filter(Gene == gene_ind[i]), aes(x = dim1, y = dim2, col = Expression)) +
    geom_point(size = 0.2, alpha = 1) +
    scale_color_gradientn(colours = cols) + # red
    facet_wrap(~Gene, scale = 'free') +
    labs(x = '', y = '', col = '') +
    theme(legend.position = 'none')
}

plot_grid(plotlist = plot_list_low, nrow = 3)
```

tiff(file = '../../outs/fig5d.tiff', width = 11, height = 7, units = 'in', res = 300)
plot_grid(plotlist = plot_list_low, nrow = 3)
dev.off()

### 20%

Upregulated in 20%

```{r}
fig5d <- ggplot(pDat_melt %>% filter(Geneset == 'Top Upregulated (20%)'),
                aes(x = dim1, y = dim2, col = Expression)) +
  geom_point(size = 0.8, alpha = 1) +
  facet_wrap(~Gene, scale = 'free', nrow = 3) +
  labs(x = 'Dimension 1', y = 'Dimension 2')
fig5d

# with cowplot::plot_grid()
plot_list_high <- list()
gene_ind <- c('PEG10', 'ANK3', 'CDC20', 'PARP1', 'GSTA3',
              'CCNA2', 'NRP2', 'CDK1', 'TOP2A', 'CCNB1')
gene_ind <- pDat_melt %>% filter(Geneset == 'Top Upregulated (20%)') %>% pull(Gene) %>% unique

for (i in 1:length(gene_ind)) {
  if (gene_ind[i] %in% c('OAS1', 'GSTA3', 'CDK1')) {
    cols <- c('#919191', '#919191', '#919191', # greys
            '#fc4e2a', # /orange
           '#e31a1c', '#bd0026') # red
  } else if (gene_ind[i] %in% c('IFIT3', 'CENPF', 'PARP1', 'TOP2A', 'KPNA2', 'NRP2')){
    cols <- c('#919191', '#919191', #GREYS 
              #'#FEB24C', #YELLOW
              '#fc4e2a', #ORANGE
              '#bd0026', '#bd0026') #red
  } else if(gene_ind[i] %in% c('DLGAP5', 'OXCT1', 'CENPF', 'KIF20A', 'NCAPH')){
    cols <-  c('#919191', '#919191','#919191', '#919191',#GREYS 
             # '#FEB24C', #YELLOW
              '#fc4e2a', #ORAGNE
              '#bd0026', '#bd0026', '#bd0026', '#bd0026') #red
  } else {
    cols <-  c('#919191', '#919191',#GREYS 
              #'#FEB24C', 
              '#fc4e2a', #ORANGE
              '#bd0026', '#bd0026', '#bd0026', '#bd0026') #red
  }
  
  plot_list_high[[i]] <-
    ggplot(pDat_melt %>% filter(Gene == gene_ind[i]), aes(x = dim1, y = dim2, col = Expression)) +
    geom_point(size = 0.2, alpha = 1) +
    scale_color_gradientn(colours = cols) + 
    facet_wrap(~Gene, scale = 'free') +
    labs(x = '', y = '', col = '') +
    theme(legend.position = 'none')
}

plot_grid(plotlist = plot_list_high, nrow = 3)
```

tiff(file = '../../outs/fig5e.tiff',  width = 11, height = 7, units = 'in', res = 300)
plot_grid(plotlist = plot_list_high, nrow = 3)
dev.off()

## 3.4 Slingshot

Here I try Slingshot because it can take a reduced dimensional matrix of the cells

```{r}
sce <- slingshot(data = troph@reductions$umap@cell.embeddings, 
                 clusterLabels = troph$celltype)

line <- getLineages(troph@reductions$umap@cell.embeddings, 
                    clusterLabels = troph$celltype, 
                    start.clus = 'VCT')
crv1 <- getCurves(line)

plot(troph@reductions$umap@cell.embeddings[,1:2], 
     col = brewer.pal(9,"Set1")[as.numeric(as.factor(troph$celltype))], asp = 1, pch = 16)
lines(line, lwd = 3, col = 'black')

plot(troph@reductions$umap@cell.embeddings[,1:2], 
     col = brewer.pal(9,"Set1")[as.numeric(as.factor(troph$celltype))], asp = 1, pch = 16)
lines(crv1, lwd = 3, col = 'black')

pDat <- troph@meta.data %>% 
  as_tibble %>% 
  bind_cols(as_tibble(sce@reducedDim)) %>% 
  bind_cols(as_tibble(troph@reductions$umap@cell.embeddings)) %>%
  mutate(pseudotime = slingPseudotime(sce)[,1])

ggplot(pDat, aes(x = UMAP_1, y = UMAP_2, col = pseudotime)) +
  geom_point()
ggplot(pDat, aes(x = celltype, y = pseudotime, col = celltype)) +
  geom_boxplot()
```